### PR TITLE
Fix dirbrowser navigation on Windows (#1334)

### PR DESCRIPTION
**Pull request by Kingdread**
Issue #1334

The problem was that there were too few slashes. On Linux, absolute
paths start with /, so

    file:// + /home

gives file:///home, which is a valid path. On windows however, absolute
paths start with a drive letter, so

    file:// + C:/Users

gives file://C:/Users, which is parsed as "host C, path Users", which is
why it could be written as file://c/Users (strip out the empty "port"),
giving us an invalid path.

The solution is to add the third slash in the template, and strip the
redundant slash on unix systems.

Additionally, this fixes a bug where navigating from '/home/' to the
parent directory would give '/home' instead of '/'

---
This change is [Reviewable](https://reviewable.io/reviews/the-compiler/qutebrowser/1360)

### CODE REVIEW

File: qutebrowser/browser/network/filescheme.py [hunk 1]
```diff
@@ -57,6 +57,14 @@ def is_root(directory):
...
         Whether the directory is a root directory or not.
     """
+    # If you're curious as why this works:
+    # dirname('/') = '/'
+    # dirname('/home') = '/'
+    # dirname('/home/') = '/home'
+    # dirname('/home/foo') = '/home'
+    # basically, for files (no trailing slash) it removes the file part, and
+    # for directories, it removes the trailing slash, so the only way for this
+    # to be equal is if the directory is the root directory.
```
Comment (The-Compiler) [inline qutebrowser/browser/network/filescheme.py]: Good idea to document this, as it isn't obvious at all.

File: qutebrowser/browser/network/filescheme.py [hunk 2]
```diff
@@ -77,14 +85,14 @@ def dirbrowser_html(path):
     if is_root(path):
         parent = None
     else:
-        parent = os.path.dirname(path)
+        parent = os.path.normpath(os.path.join(path, '..'))
```
Comment (The-Compiler) [inline qutebrowser/browser/network/filescheme.py]: Use `os.pardir` instead of `'..'` here - even though that's currently `..` everywhere, it makes the intent more clear

File: qutebrowser/utils/jinja.py [hunk 3]
```diff
@@ -71,5 +71,16 @@ def resource_url(path):
...
+
+    Arguments:
+        path: The absolute path to the local file
+    """
+    return QUrl.fromLocalFile(path).toString(QUrl.FullyEncoded)
+
+
 env = jinja2.Environment(loader=Loader('html'), autoescape=_guess_autoescape)
 env.globals['resource_url'] = resource_url
+env.globals['file_url'] = file_url
```
Comment (The-Compiler) [inline qutebrowser/utils/jinja.py]: Already looking forward to all the new crash reports because people updated qutebrowser (and thus the `.html` on disk) but didn't restart it (thus `file_url` is missing). The same happened with `resource_url`, but I don't think there's much I can do :laughing:

**Comment by The-Compiler** [comment]
Thanks! LGTM apart from the small nitpick.

It'd be really useful to have some more tests for those two cases:

- `C:/Users` vs `c/users` on Windows
- Moving to the parent from e.g. `/home/` on Linux

Probably either in `tests/unit/browser/network/test_filescheme.py`, or as an end-to-end test (maybe written in Python like `tests/integration/test_hints_html.py` so it can use `tmpdir` easily).

I haven't looked at it in detail yet, so if you want to, please do! If not, tell me and I'll add it to my todo list :wink:

File: qutebrowser/utils/jinja.py (hunk 3 above)
Comment (The-Compiler) [inline qutebrowser/utils/jinja.py]: On a second thought, I guess we could simply handle `jinja2.exceptions.UndefinedError` in `render`, and return a static HTML which tells the user to upgrade. I'll take a look at that later.

**Comment by Kingdread** [comment]
After fighting with segfaults on Windows, I've finally tested it and the test fails without the fix applied (the `c/Users` thing) and passes with the new method.

File: qutebrowser/utils/jinja.py (hunk 3 above)
Comment (Kingdread) [inline qutebrowser/utils/jinja.py]: Good idea, probably good to merge a fix for #1362 first then :wink:

File: tests/integration/quteprocess.py [hunk 4]
```diff
@@ -359,6 +359,18 @@ def open_path(self, path, *, new_tab=False, new_window=False, port=None,
         else:
             self.send_cmd(':open ' + url)
 
+    def open_url(self, url, *, new_tab=False, new_window=False):
```
Comment (The-Compiler) [inline tests/integration/quteprocess.py]: Can you adjust `open_path` to simply call `path_to_url` and then `open_url` instead of duplicating the code?

Comment (Kingdread) [inline tests/integration/quteprocess.py]: Yes, that was an oversight on my part

**Comment by Kingdread** [comment]
Rebased on the current master to include the jinja render changes (#1362).

**Comment by The-Compiler** [comment]
Thanks! I'll do another review and/or merge tomorrow :wink:

File: tests/integration/test_dirbrowser.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+@pytest.mark.parametrize('folder', DirLayout.layout_folders())
+def test_enter_folder(dir_layout, quteproc, folder):
+    quteproc.open_url(dir_layout.file_url())
+    # Use Javascript and XPath to click the link that has the folder name as
+    # text.
+    quteproc.send_cmd(
+        ':jseval document.evaluate(\'//a[text()="{}"]\', document, null, '
+        'XPathResult.ANY_TYPE, null).iterateNext().click()'
+        .format(folder)
+    )
```
Comment (The-Compiler) [inline tests/integration/test_dirbrowser.py]: Hmm, maybe that should be a `quteproc.click_element(text=...)` method or so? Do you think it could be useful for other tests too?

File: tests/integration/test_dirbrowser.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+        title_prefix = 'Browse directory: '
+        # Strip off the title prefix to obtain the path of the folder that
+        # we're browsing
+        path = soup.title.string[len(title_prefix):]
+        path = os.path.normpath(path)
+
+        container = soup('div', id='dirbrowserContainer')[0]
+
+        parent_elem = container('ul', class_='parent')
+        if len(parent_elem) == 0:
```
Comment (The-Compiler) [inline tests/integration/test_dirbrowser.py]: nitpick: I think this can simply be `if not parent_elem:` as it should evaluate to false? Though I'm not sure what to expect with beautifulsoup :laughing:

File: tests/integration/test_dirbrowser.py [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+
+    def path(self, *parts):
+        """Return the path to the given file inside the layout folder."""
+        return os.path.normpath(str(self.layout.join(*parts)))
+
+    def base_path(self):
+        """Return the path of the base temporary folder."""
+        return os.path.normpath(str(self.base))
+
+    def parse(self, quteproc):
```
Comment (The-Compiler) [inline tests/integration/test_dirbrowser.py]: Hmm, having `parse`/`Item`/`Parsed` in here seems kind of out-of-place - what about having those as top-level functions/classes in this file instead?

File: tests/integration/test_dirbrowser.py (hunk 5 above)
Comment (Kingdread) [inline tests/integration/test_dirbrowser.py]: Could be useful, but I'd definitely make it more robust then, add some tests for it :wink: and specify what will happen with zero/ambiguous elements (click all of them? click the first? raise an error?)

File: tests/integration/test_dirbrowser.py (hunk 6 above)
Comment (Kingdread) [inline tests/integration/test_dirbrowser.py]: I took it from the unit test, so I'm not sure either - but I'll try it :smile:

File: tests/integration/test_dirbrowser.py (hunk 7 above)
Comment (Kingdread) [inline tests/integration/test_dirbrowser.py]: Sounds good, it landed there because I thought I'd need to pass the temp dir myself when getting the html content, but as it turns out, quteproc already has such a helper function :wink:

File: tests/integration/test_dirbrowser.py (hunk 5 above)
Comment (The-Compiler) [inline tests/integration/test_dirbrowser.py]: Would be nice. You should probably use `javascript_escape` from `qutebrowser/browser/webelem.py` too.

I'd say raise an exception when there's != 1 element.

**Comment by Kingdread** [comment]
I've added the `click_element` method to `quteproc`.

I realize that the result from `_xpath_escape` could be more efficient (keep together same quotes), but always splitting at a quote made the function pretty simple to write, so I'd prefer to keep the implementation easy (even if the output is not optimal).

**Comment by The-Compiler** [comment]
Great work, thank you!
